qtfred/src/mission/dialogs/CampaignEditorDialogModel.h

```diff
@@ -0,0 +1,254 @@
+#pragma once
+
+#include "mission/dialogs/AbstractDialogModel.h"
+#include "ui/dialogs/CampaignEditorDialog.h"
+#include "AssociatedPlainTextDocument.h"
+#include "CheckedDataListModel.h"
+#include <mission/missioncampaign.h>
+#include <ui/widgets/sexp_tree.h>
+
+#include <QTextDocument>
+#include <QPlainTextEdit>
+
+namespace fso {
+namespace fred {
+namespace dialogs {
+
+static const QString qstrEmpty{};
+
+class CampaignEditorDialog;
+
+class CampaignEditorDialogModel : public AbstractDialogModel, public SexpTreeEditorInterface
+{
+	struct CampaignMissionData;
+	struct CampaignBranchData;
+	Q_OBJECT
+
+public:
+	CampaignEditorDialogModel(CampaignEditorDialog *parent, EditorViewport *viewport, const QString &file = "", const QString& newCampaignType = "");
+	~CampaignEditorDialogModel() override = default;
+	void supplySubModels(QListView &ships, QListView &weps, QListView &missions, QPlainTextEdit &descr);
+
+// AbstractDialogModel
+	bool apply() override {	return saveTo(campaignFile); }
+	void reject() override {} // nothing to do if the dialog is created each time it's opened
+
+// SexpTreeEditorInterface
+	QStringList getMissionGoals(const QString& reference_name) override;
+	bool hasDefaultGoal(int) override {return true;}
+
+	QStringList getMissionEvents(const QString& reference_name) override;
+	bool hasDefaultEvent(int) override {return true;}
+
+	QStringList getMissionNames() override;
+	bool hasDefaultMissionName() override {return true;}
+
+	bool requireCampaignOperators() const override {return true;}
+
+	QList<QAction *> getContextMenuExtras(QObject *parent = nullptr) override;
+
+// Model state getters
+	inline bool isFileLoaded() const { return ! campaignFile.isEmpty(); }
+
+	inline const QString& getCampaignName() const { return campaignName; }
+	inline bool getCampaignTechReset() const { return campaignTechReset; }
+	int getCampaignNumPlayers() const;
+
+	inline bool isCurMnSelected() const { return mnData_it; }
```

Comment:
I'm using mn/Mn and br/Br as abbreviations for mission and branch throughout. I can add a comment to that effect, renaming would cause some really long function names

Reply:
Gotcha. Yeah, a single comment explaining the naming convention would be nice.